python/lsst/ts/m2/tcp_client.py

```diff
@@ -90,30 +98,39 @@ def __init__(
         self.timeout_in_second = timeout_in_second
 
         # Sequence ID generator
-        self._sequence_id_generator = salobj.index_generator()
+        if sequence_generator is None:
+            self._sequence_id_generator = salobj.index_generator()
+        else:
+            self._sequence_id_generator = sequence_generator
         self.last_sequence_id = -1
 
         self.queue = asyncio.Queue(maxsize=int(maxsize_queue))
 
         # Monitor loop task (asyncio.Future)
         self._monitor_loop_task = make_done_future()
 
-    async def connect(self, connect_retry_interval=1.0):
+    async def connect(self, connect_retry_interval=1.0, timeout=10.0):
         """Connect to the server.
 
         Parameters
         ----------
         connect_retry_interval : `float`, optional
             How long to wait before trying to reconnect when connection fails.
             (default is 1.0)
+        timeout : `float`, optional
+            Timeout in second. This value should be larger than the
+            connect_retry_interval. (default is 10.0)
 
-        Notes
-        -----
-        This will wait forever for a connection.
+        Raises
+        ------
+        asyncio.TimeoutError
+            Connection timeout.
         """
 
-        self.log.info("Open the connection.")
+        self.log.info("Try to open the connection.")
 
+        retry_times_max = timeout // connect_retry_interval
+        retry_times = 0
```

Comment:
Do you realize this may actually take longer than `timeout`? The timeout is not taking into account the time it takes to execute `asyncio.open_connection`. Also, what if `asyncio.open_connection` hangs for ever?  This would create a deadlock here. 

Consider refactoring this code to improve how the timeout is handled..

Reply:
`asyncio.open_connection()` is a non-block call and return the result or raise the error immediately.